src/utils.rs

```diff
@@ -6,85 +6,48 @@ use core::ptr::NonNull;
 
 use alloc::vec::Vec;
 
-/// OS memory casting macros.
-mod memory_casting {
-    /// Cast a cached address to a uncached address.
-    /// Example: 0x8xxxxxxx -> 0xCxxxxxxx
-    #[macro_export]
-    macro_rules! mem_cached_to_uncached {
-        ( $x:expr ) => {{
-            use core::ffi::c_void;
-
-            (($x as u32) + ($crate::ffi::SYS_BASE_UNCACHED - $crate::ffi::SYS_BASE_CACHED))
-                as *mut c_void
-        }};
-    }
+/// OS memory pointer casting.
+/// 
+/// For more information, refer to [Memory map](https://wiibrew.org/wiki/Memory_Map).
+pub mod mem {
+    use crate::ffi;
 
-    /// Cast a cached address to a physical address.
-    /// Example: 0x8xxxxxxx -> 0x0xxxxxxx
-    #[macro_export]
-    macro_rules! mem_cached_to_physical {
-        ( $x:expr ) => {{
-            use core::ffi::c_void;
+    pub const BASE_CACHED: usize = ffi::SYS_BASE_CACHED as _;
+    pub const BASE_UNCACHED: usize = ffi::SYS_BASE_UNCACHED as _;
 
-            (($x as u32) - $crate::ffi::SYS_BASE_CACHED) as *mut c_void
-        }};
+    /// Cast a cached address to an uncached address.
+    /// Example: 0x8xxxxxxx -> 0xCxxxxxxx
+    #[inline]
+    pub fn cached_to_uncached(x: usize) -> usize {
+        physical_to_uncached(virtual_to_physical(x))
```

Comment:
This could backfire on people who don't know which address space they are in.

people passing invalid usizes such at 0xC0000 and and 0x0000000 related addrs could cause issues